### ISSUE TITLE
Mark register as fixed

### ISSUE THREAD

**Issue by Leseratte10** [Vector35/binaryninja-api#1149]
I am analyzing a binary for the PowerPC EABI architecture. 

The PowerPC has two fixed registers, r2 and r13, which are set once on program start, and are never modified again. Those registers point to two small-data-areas (r2 being read-only, r13 being read-write), so commonly used variables are stored around the memory addresses in these two registers, so they can be accessed with a single `lwz r0, X(r13)` with X being the offset, without having to waste another instruction on loading the other address half. 

Is there any way I can tell Binary Ninja that that is the case? 

The code contains lines like `stw r3, -27648(r13)`, storing data to a variable inside this small data area, but Binary Ninja doesn't recognize that. I cannot right-click and name that variable and have the variable name show up anywhere else this address (r13 - 27648) shows up, like I could if there were a "lis r13, X" the line before ...

**Comment by toolCHAINZ** [comment]
This is something that might be fixed if https://github.com/Vector35/binaryninja-api/issues/281 is implemented.

Alternatively, you could write an architecture hook (https://api.binary.ninja/binaryninja.architecture.ArchitectureHook.html) for the MIPS module that overrides the lifting of those instructions to return a constant value instead of the register contents. Which might be a bit messy, but should get the job done.

**Comment by joshwatson** [comment]
I think what you actually want is to implement an ArchitectureHook that adds those registers to the `global_regs` property

**Comment by toolCHAINZ** [comment]
@joshwatson are there any good examples/documentation for `global_regs`? I was not aware of that but it sounds interesting/useful. Will it treat those registers like stack pointers or something?

**Comment by Leseratte10** [comment]
I opened a python shell within Binary Ninja, and entered the following: 
```
a = Architecture["ppc"]
a.global_regs = [2, 13]
```
is that correct? Or, how else do I set this global_regs property? After re-analyzing the binary, instructions like `addi r4, r13, -0x7ba4` still don't let me name the variable at r13-0x7ba4.

**Comment by joshwatson** [comment]
Well there's a couple issues with that. First, `global_regs` is just a local for the `__init__` method if you assign it like that. What you would want to do is something like move that to be a class member, and then you'd implement `ppcsdrHook._get_global_registers` that returns that list, and then set `ppcsdrHook._cb.getGlobalRegisters` to that method. However, I'm not _completely_ sure that an ArchitectureHook can override that method, based on the source for that class. One of the Vector35 people would have to confirm.

**Comment by rssor** [comment]
`ArchitectureHook` should be capable of overriding every single method as far as the core is concerned -- if it can't be done in the Python API that would be a bug in the bindings.

**Comment by Leseratte10** [comment]
Trying to do that gives me the error "ppcsdrHook has no attribute _cb". 

Aren't there any examples for such a hook? Am I the first one who wants to modify the behaviour of an architecture?

```
from binaryninja import *

class ppcsdrHook(ArchitectureHook):

	def _get_global_registers(): 
		return [2, 13]
	def __init__(self, base_arch):
		super(ppcsdrHook, self).__init__(base_arch)
		ppcsdrHook._cb.getGlobalRegisters = self._cb.getGlobalRegisters.__class__(self._get_global_registers)



a = ppcsdrHook(Architecture['ppc'])
a.register()
```

**Comment by joshwatson** [comment]
No, there have been a few. There's one in the python/examples directory

**Comment by Leseratte10** [comment]
Yeah but that is only a basic one and doesn't use the callbacks (which don't work for me).

**Comment by joshwatson** [comment]
It may be that the ArchitectureHook doesn't implement a hook for global registers.

**Comment by rssor** [comment]
Global registers in the Python bindings don't require you to implement a callback, you'd just provide a `global_regs = []` member in the class definition, as that's where the callbacks look for them.

**Comment by Leseratte10** [comment]
OK, but how exactly does that work? I tried the following, the plugin loads without errors ("Loaded python plugin 'test'"), but references to r13 are still not recognized as symbols: 

```
from binaryninja.architecture import Architecture, ArchitectureHook
from binaryninja import *

class ppcsdrHook(ArchitectureHook):
	global_regs = [2, 13]

a = ppcsdrHook(Architecture['ppc'])
a.register()
```

Am I doing something wrong? If so, can you tell me what I need to change?

**Comment by joshwatson** [comment]
Use 58 instead of 13 and 47 instead of 2

**Comment by Leseratte10** [comment]
Want me to provide a sample binary?

**Comment by rssor** [comment]
Sure -- if you can't post it here, feel free to send it to me directly on slack (rss) or to the support email (binaryninja@vector35.com).

**Comment by Leseratte10** [comment]
Thanks for your help. I sent you a message on slack.

**Comment by Leseratte10** [comment]
Is there any estimate when this will be fixed? You mentioned on Slack you were going to look into it shortly after the last release, which was at the end of September...

I'd really like to be able to assign variable names to these r2 and r13 references so BN recognizes them when used elsewhere and I don't have to manually write down variable names and the corresponding r13 offsets ...

**Comment by Leseratte10** [comment]
Sorry for bringing this up again but is there any ETA for when this will be fixed? It is kinda annoying not being able to assign variable names to these small data areas. I updated to 1.1.1460-dev and it still doesn't recognize r13 and r2 as being fixed, even though there is only one part in the whole binary that actually sets r13 and r2.

**Comment by Leseratte10** [comment]
It looks like the bug has been fixed in 1.1.1461-dev, references to r2 and r13 now correctly display the RAM address and I can assign variable names to them. Thanks a lot for fixing this!
